### ISSUE TITLE
[dataio] Undefined behavior in dataio/icetray (Trac #2216)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2208]
I can elicit undefined behavior in dataio unit tests by the following:

1. Run a_nocompression.py
2. Run i_adds_mutineer.py
3. Run j_fatals_reading_mutineer.py with the addition of 'from icecube.test_unregistered import UnregisteredTrack'.  I've pasted a single python file below that elicits this behavior.

The test randomly passes/fails.  I can only reproduce this behavior on Mac OSX, and my test setup included Python 2.7.14 and LLVM-9.0.0.

#!/usr/bin/env python

from I3Tray import *

from os.path import expandvars

import os
import sys

from icecube import dataclasses 
from icecube import phys_services 
from icecube import dataio 

from icecube.test_unregistered import UnregisteredTrack

#
# This sets up a bunch of paths of files and stuff.  Nice to have a
# real scripting language at one's disposal for this kind of thing.
#
testdata = expandvars("$I3_TESTDATA")
runfile = testdata + "/2007data/2007_I3Only_Run109732_Nch20.i3.gz"

tray = I3Tray()

tray.AddModule("I3Reader",
               Filename = runfile,
               SkipKeys = ["I3PfFilterMask","CalibratedFADC","CalibratedATWD"])

# This file is super old
tray.AddModule("QConverter")
tray.AddModule(lambda fr: False) # Drop all existing P-frames

#
# Make the Q frames into P
#
tray.AddModule("I3NullSplitter")

#
# And this is the magic writer.  We will make it work harder later.
#
tray.AddModule("I3Writer", filename =  "pass1.i3")

#
# Here we specify how many frames to process, or we can omit the
# argument to Execute() and the the tray will run until a module tells
# it to stop (via RequestSuspension()).  We'll do a few frames so
# there's a chunk of data involved.
#
tray.Execute(15)

tray = I3Tray()
tray.AddModule("I3Reader", Filename="pass1.i3")
tray.AddModule(lambda frame : \
               frame.Put("mutineer", UnregisteredTrack()))
tray.AddModule("Dump")
tray.AddModule("I3Writer", Filename = "hasmutineer.i3.gz")

tray.Execute()

tray = I3Tray()

# by default the reader will log_fatal if it can't deserialize something
tray.AddModule("I3Reader","reader", Filename=expandvars("hasmutineer.i3.gz"))

# this guy actually does the get and causes the error
tray.AddModule("Get", "getter")(
    ("Keys", ["mutineer"])
    )

tray.AddModule("Dump","dump")

try:
    tray.Execute()
    

except:
    sys.exit(0)
else:
    print("***\n***\n*** Failure! Script didn't throw as it should have.\n***\n***\n***\n")
    sys.exit(1) # ought to die, shouldn't get here

Migrated from https://code.icecube.wisc.edu/ticket/2216
```json
{
    "status": "closed", 
    "changetime": "2019-03-19T15:14:54", 
    "description": "I can elicit undefined behavior in dataio unit tests by the following:\n\n1. Run a_nocompression.py\n2. Run i_adds_mutineer.py\n3. Run j_fatals_reading_mutineer.py with the addition of 'from icecube.test_unregistered import UnregisteredTrack'.  I've pasted a single python file below that elicits this behavior.\n\nThe test randomly passes/fails.  I can only reproduce this behavior on Mac OSX, and my test setup included Python 2.7.14 and LLVM-9.0.0.\n\n\n#!/usr/bin/env python\n\nfrom I3Tray import *\n\nfrom os.path import expandvars\n\nimport os\nimport sys\n\nfrom icecube import dataclasses \nfrom icecube import phys_services \nfrom icecube import dataio \n\nfrom icecube.test_unregistered import UnregisteredTrack\n\n#\n# This sets up a bunch of paths of files and stuff.  Nice to have a\n# real scripting language at one's disposal for this kind of thing.\n#\ntestdata = expandvars(\"$I3_TESTDATA\")\nrunfile = testdata + \"/2007data/2007_I3Only_Run109732_Nch20.i3.gz\"\n\ntray = I3Tray()\n\ntray.AddModule(\"I3Reader\",\n               Filename = runfile,\n               SkipKeys = [\"I3PfFilterMask\",\"CalibratedFADC\",\"CalibratedATWD\"])\n\n# This file is super old\ntray.AddModule(\"QConverter\")\ntray.AddModule(lambda fr: False) # Drop all existing P-frames\n\n#\n# Make the Q frames into P\n#\ntray.AddModule(\"I3NullSplitter\")\n\n#\n# And this is the magic writer.  We will make it work harder later.\n#\ntray.AddModule(\"I3Writer\", filename =  \"pass1.i3\")\n\n#\n# Here we specify how many frames to process, or we can omit the\n# argument to Execute() and the the tray will run until a module tells\n# it to stop (via RequestSuspension()).  We'll do a few frames so\n# there's a chunk of data involved.\n#\ntray.Execute(15)\n\ntray = I3Tray()\ntray.AddModule(\"I3Reader\", Filename=\"pass1.i3\")\ntray.AddModule(lambda frame : \\\n               frame.Put(\"mutineer\", UnregisteredTrack()))\ntray.AddModule(\"Dump\")\ntray.AddModule(\"I3Writer\", Filename = \"hasmutineer.i3.gz\")\n\ntray.Execute()\n\ntray = I3Tray()\n\n# by default the reader will log_fatal if it can't deserialize something\ntray.AddModule(\"I3Reader\",\"reader\", Filename=expandvars(\"hasmutineer.i3.gz\"))\n\n# this guy actually does the get and causes the error\ntray.AddModule(\"Get\", \"getter\")(\n    (\"Keys\", [\"mutineer\"])\n    )\n\ntray.AddModule(\"Dump\",\"dump\")\n\ntry:\n    tray.Execute()\n    \n\nexcept:\n    sys.exit(0)\nelse:\n    print(\"***\\n***\\n*** Failure! Script didn't throw as it should have.\\n***\\n***\\n***\\n\")\n    sys.exit(1) # ought to die, shouldn't get here", 
    "reporter": "jbraun", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1553008494085899", 
    "component": "combo core", 
    "summary": "[dataio] Undefined behavior in dataio/icetray", 
    "priority": "major", 
    "keywords": "bug icetray dataio", 
    "time": "2018-12-04T16:15:08", 
    "milestone": "Vernal Equinox 2019", 
    "owner": "olivas", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20190111T21:58:21: 
* olivas changed status from "new" to "accepted"
* olivas changed owner from "" to "olivas"

**Comment by icecube-trac** [comment]
Trac update at 20190319T11:22:32: olivas changed attachment from "" to "undefined_behavior.py"

**Comment by icecube-trac** [comment]
Trac update at 20190319T11:31:03: olivas commented:

> The pasted script needed a couple tweaks, but attached is a working script.  I also get undefined behavior on Ubuntu 18.04/python 2.7.15rc1/gcc7.3.0.  Sometimes this fails, sometimes it doesn't.

**Comment by icecube-trac** [comment]
Trac update at 20190319T13:20:30: olivas commented:

> Got this to fail once after many passes with a debug build as well.

**Comment by icecube-trac** [comment]
Trac update at 20190319T13:54:38: olivas commented:

> Seems the unregistered class is the JEBEventInfo and not the mutineer, which brings up more questions than it answers...

**Comment by icecube-trac** [comment]
Trac update at 20190319T13:57:03: 
* olivas changed _comment0 from:

> ERROR (I3Module): !fop = 0 (Get.cxx:94 in void Get::impl(I3FramePtr))
> ERROR (I3Module): frame->Has(mutineer) = 1 (Get.cxx:95 in void Get::impl(I3FramePtr))
> 
> I3Frame happily creates a shared pointer for the mutineer.

to:

> 1553003838963176


* olivas commented:

> ERROR (I3Module): !fop = 0 (Get.cxx:94 in void Get::impl(I3FramePtr))
> ERROR (I3Module): frame->Has(mutineer) = 1 (Get.cxx:95 in void Get::impl(I3FramePtr))
> 
> I3Frame::Get happily creates a shared pointer for the mutineer.

**Comment by icecube-trac** [comment]
Trac update at 20190319T14:43:41: olivas commented:

> When this test passes it's because it fails immediately on JEBEventInfo (at frame 6), which is actually unregistered.  When it doesn't throw (and the test therefore fails) it will happily create about 20 mutineer objects.

**Comment by icecube-trac** [comment]
Trac update at 20190319T14:50:14: olivas commented:

> I believe I found the source of undefined behavior, but it still doesn't explain why mutineers are happily created.
> 
> The get_all_ member variable (of the Get module) is not initialized when the size of the Keys parameter is non-zero. So it tries to read everything in the frame sometimes (and sometimes not) and throws (as is should) in JEBEventInfo, but that's not where we thought it was going to throw.  We expected it to throw when it tried to deserialize a mutineer, create solely for this purpose. 
> 
> I wonder for how long now we've been happily creating objects we couldn't possibly create correctly because they're unregistered.

**Comment by icecube-trac** [comment]
Trac update at 20190319T15:13:56: olivas commented:

> Ok.  So the uninitialized get_all_ in the Get module was in fact the source of the undefined behavior.
> 
> Attached are the two scripts split between write and read.  Can't do this in one script since I don't think python allows you to 'unload' a library (there might be a hacky way to do this, but beyond the scope of this ticket).
> 
> Now when you read hasmutineer.i3.gz with the reader script, making sure *not* to import UnregisteredTrack, the tray throws and the test passes as expected.

**Comment by icecube-trac** [comment]
Trac update at 20190319T15:14:16: olivas changed attachment from "" to "undefined_behavior_writer.py"

**Comment by icecube-trac** [comment]
Trac update at 20190319T15:14:25: olivas changed attachment from "" to "undefined_behavior_reader.py"

**Comment by icecube-trac** [comment]
Trac update at 20190319T15:14:54: 
* olivas commented:

> In changeset 172148:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="172148"
> initialize all your member variables.  vectors are initialized empty, so we're good.  fixes #2208.
> ```


* olivas changed resolution from "" to "fixed"
* olivas changed status from "accepted" to "closed"
